File: packages/tracelet_platform_interface/test/algorithms/battery_budget_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tracelet_platform_interface/tracelet_platform_interface.dart';

void main() async {
  await RustLib.init();
  group('BatteryBudgetEngine', () {
    test('constructor sets defaults', () {
      final engine = BatteryBudgetEngine();

      expect(engine.targetBudgetPerHour, 3.0);
      expect(engine.distanceFilter, 10.0);
      expect(engine.accuracyIndex, 0);
      expect(engine.periodicInterval, isNull);
    });

    test('constructor accepts custom initial values', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 50,
        initialAccuracyIndex: 2,
        initialPeriodicInterval: 300,
      );

      expect(engine.targetBudgetPerHour, 5.0);
      expect(engine.distanceFilter, 50.0);
      expect(engine.accuracyIndex, 2);
      expect(engine.periodicInterval, 300);
    });

    test('clamps accuracy index to 0-4 range', () {
      final tooHigh = BatteryBudgetEngine(initialAccuracyIndex: 10);
      expect(tooHigh.accuracyIndex, 4);

      final tooLow = BatteryBudgetEngine(initialAccuracyIndex: -5);
      expect(tooLow.accuracyIndex, 0);
    });

    test('first processSample returns null (baseline)', () {
      final engine = BatteryBudgetEngine();

      final result = engine.processSample(0.95);
      expect(result, isNull);
    });

    test('second processSample within 60s returns null', () {
      final engine = BatteryBudgetEngine();

      // First call sets baseline
      engine.processSample(0.95);

      // Immediate second call — too soon (< 60 seconds)
      final result = engine.processSample(0.90);
      expect(result, isNull);
    });

    test('reset allows re-establishing baseline', () {
      final engine = BatteryBudgetEngine();

      engine.processSample(0.95);
      engine.reset();

      // After reset, next call should return null (new baseline)
      final result = engine.processSample(0.90);
      expect(result, isNull);
    });

    test('reset does not change configuration', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 50,
        initialAccuracyIndex: 2,
      );

      engine.processSample(0.95);
      engine.reset();

      expect(engine.targetBudgetPerHour, 5.0);
      expect(engine.distanceFilter, 50.0);
      expect(engine.accuracyIndex, 2);
    });
  });

  group('BudgetAdjustmentEvent', () {
    test('constructor stores all fields', () {
      const event = BudgetAdjustmentEvent(
        currentBatteryDrain: 4.5,
        targetBudget: 3,
        newDistanceFilter: 50,
        newDesiredAccuracy: 2,
        newPeriodicInterval: 600,
      );

      expect(event.currentBatteryDrain, 4.5);
      expect(event.targetBudget, 3.0);
      expect(event.newDistanceFilter, 50.0);
      expect(event.newDesiredAccuracy, 2);
      expect(event.newPeriodicInterval, 600);
    });

    test('newPeriodicInterval defaults to null', () {
      const event = BudgetAdjustmentEvent(
        currentBatteryDrain: 2,
        targetBudget: 3,
        newDistanceFilter: 10,
        newDesiredAccuracy: 0,
      );

      expect(event.newPeriodicInterval, isNull);
    });
  });

  group('BatteryBudgetEngine — internal state', () {
    test('initial distanceFilter matches constructor param', () {
      final engine = BatteryBudgetEngine(initialDistanceFilter: 100);
      expect(engine.distanceFilter, 100.0);
    });

    test('initial periodicInterval matches constructor param', () {
      final engine = BatteryBudgetEngine(initialPeriodicInterval: 300);
      expect(engine.periodicInterval, 300);
    });

    test('periodicInterval is null when not provided', () {
      final engine = BatteryBudgetEngine();
      expect(engine.periodicInterval, isNull);
    });

    test('accuracy index clamped to minimum 0', () {
      final engine = BatteryBudgetEngine(initialAccuracyIndex: -100);
      expect(engine.accuracyIndex, 0);
    });

    test('accuracy index clamped to maximum 4', () {
      final engine = BatteryBudgetEngine(initialAccuracyIndex: 100);
      expect(engine.accuracyIndex, 4);
    });

    test('reset preserves distance filter and accuracy', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 200,
        initialAccuracyIndex: 3,
        initialPeriodicInterval: 600,
      );

      // Establish baseline then reset
      engine.processSample(0.95);
      engine.reset();

      // Configuration should be preserved
      expect(engine.distanceFilter, 200.0);
      expect(engine.accuracyIndex, 3);
      expect(engine.periodicInterval, 600);

      // Next processSample should return null (new baseline)
      expect(engine.processSample(0.90), isNull);
    });

    test('charging (battery increase) returns null', () {
      final engine = BatteryBudgetEngine();

      // The engine uses DateTime.now() internally so we can only test
      // the first-sample baseline behavior directly. Charging detection
      // requires elapsed time > 60s which can't be simulated without
      // injecting a clock.
      engine.processSample(0.50);
      // Immediately calling with higher level returns null (too soon)
      expect(engine.processSample(0.60), isNull);
    });
  });

  /// The ladder replaced an unbounded multiplier in #393/#396, and these tests
  /// replaced the ones that described it.
  ///
  /// Every assertion here used to encode the defect rather than a requirement:
  /// a single one-hour window was enough to throttle, `distanceFilter` was
  /// multiplied by 1.5 and clamped up to a floor of 10 m regardless of what the
  /// app had configured, and accuracy was coarsened on the first over-budget
  /// reading. The field failure that prompted the rewrite was a device that had
  /// simply crossed one 5 % battery reporting step.
  group('BatteryBudgetEngine — the throttle ladder (clock injection)', () {
    late DateTime fakeNow;
    late BatteryBudgetEngine engine;

    setUp(() {
      fakeNow = DateTime(2024, 1, 1, 12);
    });

    BatteryBudgetEngine createEngine({
      double targetBudgetPerHour = 3.0,
      double initialDistanceFilter = 10.0,
      int initialAccuracyIndex = 0,
      int? initialPeriodicInterval,
    }) {
      return BatteryBudgetEngine(
        targetBudgetPerHour: targetBudgetPerHour,
        initialDistanceFilter: initialDistanceFilter,
        initialAccuracyIndex: initialAccuracyIndex,
        initialPeriodicInterval: initialPeriodicInterval,
        clock: () => fakeNow,
      );
    }

    /// Two consecutive conclusive over-budget windows — the dwell one rung
    /// takes. 20 %/hr over hour-long windows, well past the 5 %/hr such a
    /// window can resolve at the default reporting granularity.
    void climbOneRung(BatteryBudgetEngine engine, {double from = 0.95}) {
      engine.processSample(from);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(from - 0.20);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(from - 0.40);
    }

    test('one conclusive window is not enough to throttle', () {
      engine = createEngine();

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // 20 %/hr against a 3 %/hr budget: conclusive, but a single window.
      final result = engine.processSample(0.75);

      expect(result, isNull, reason: 'a lone window is not a dwell');
      expect(engine.distanceFilter, 10.0);
      expect(engine.accuracyIndex, 0);
    });

    test('a drain inside the measurement resolution never throttles', () {
      engine = createEngine();

      // 4 %/hr against a 3 %/hr budget, with a window that can only resolve
      // 5 %/hr — indistinguishable from the reporting step itself.
      for (var i = 0; i < 4; i++) {
        engine.processSample(0.95 - i * 0.04);
        fakeNow = fakeNow.add(const Duration(hours: 1));
      }

      expect(engine.distanceFilter, 10.0);
      expect(engine.accuracyIndex, 0);
    });

    test('sustained heavy drain climbs one rung, cadence first', () {
      engine = createEngine();

      climbOneRung(engine);

      // Rung 1 raises the platform distance filter to its floor and leaves
      // accuracy alone: on iOS the tier below Best is 100 m, which a 15 m
      // tracking gate would then reject wholesale.
      expect(engine.distanceFilter, 10.0);
      expect(engine.accuracyIndex, 0);
    });

    test('the ladder never goes below the configured values', () {
      engine = createEngine(
        initialDistanceFilter: 250,
        initialAccuracyIndex: 2,
      );

      climbOneRung(engine);

      expect(
        engine.distanceFilter,
        250.0,
        reason: 'a configured filter wider than the rung stands',
      );
      expect(
        engine.accuracyIndex,
        2,
        reason: 'a configured tier coarser than the rung stands',
      );
    });

    test('a configured distanceFilter of 0 is never clamped up', () {
      engine = createEngine(initialDistanceFilter: 0);

      expect(engine.distanceFilter, 0.0);

      climbOneRung(engine);

      // The opt-out the old engine destroyed first: (0 * 1.5).clamp(10, 5000)
      // became 10 m, permanently, since the recovery path clamped at 10 too.
      expect(engine.distanceFilter, greaterThanOrEqualTo(0.0));
    });

    test('the accuracy index stays within 0-4 however long it throttles', () {
      engine = createEngine(targetBudgetPerHour: 1, initialAccuracyIndex: 3);

      for (var round = 0; round < 6; round++) {
        engine.processSample(0.95);
        fakeNow = fakeNow.add(const Duration(hours: 1));
        engine.processSample(0.75);
        fakeNow = fakeNow.add(const Duration(hours: 1));
      }

      expect(engine.accuracyIndex, inInclusiveRange(0, 4));
    });

    test(
      'the periodic interval is stretched, never below the configured one',
      () {
        engine = createEngine(initialPeriodicInterval: 300);

        climbOneRung(engine);

        expect(engine.periodicInterval, isNotNull);
        expect(
          engine.periodicInterval,
          greaterThanOrEqualTo(300),
          reason: 'throttling stretches cadence, it never tightens it',
        );
        expect(engine.periodicInterval, lessThanOrEqualTo(43200));
      },
    );

    test('no adjustment when the drain sits on budget', () {
      engine = createEngine();

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Exactly on budget: neither direction is conclusive.
      expect(engine.processSample(0.92), isNull);
    });

    test('no adjustment when charging', () {
      engine = createEngine();

      engine.processSample(0.50);
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Battery went UP. One window cannot move the ladder in either
      // direction, and there is nothing to lift at rung 0.
      expect(engine.processSample(0.60), isNull);
    });

    test('too-soon sample is ignored', () {
      engine = createEngine();

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(seconds: 30));

      expect(engine.processSample(0.50), isNull);
    });

    test('a throttled session comes back down when the drain does', () {
      engine = createEngine();

      climbOneRung(engine);
      final throttledFilter = engine.distanceFilter;

      // Two conclusive under-budget windows: 0.25 %/hr over four hours, where
      // the window resolves 1.25 %/hr.
      for (var i = 0; i < 2; i++) {
        fakeNow = fakeNow.add(const Duration(hours: 4));
        engine.processSample(0.50 - i * 0.01);
      }

      expect(
        engine.distanceFilter,
        lessThanOrEqualTo(throttledFilter),
        reason: 'recovery uses the same evidence bar as escalation',
      );
    });
  });
}

File: sdk/android/tracelet-sdk/src/test/kotlin/com/ikolvi/tracelet/sdk/algorithm/BatteryBudgetEngineTest.kt
package com.ikolvi.tracelet.sdk.algorithm

import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for [BatteryBudgetEngine].
 *
 * These assert the ladder's contract, which replaced an unbounded multiplier in
 * #393/#396. The tests that stood here before asserted the multiplier's
 * behaviour, and every one of them described the defect rather than the
 * requirement: a single five-minute window was enough to throttle, a configured
 * distance filter was clamped up to 10 m, and accuracy was coarsened on the
 * first over-budget reading. The field failure was a device that had simply
 * crossed one 5 % battery reporting step.
 */
class BatteryBudgetEngineTest {

    private companion object {
        const val MINUTE_MS = 60_000L

        /** Android's own reporting step; the engine is told this at construction. */
        const val QUANTUM = 1.0
    }

    private lateinit var engine: BatteryBudgetEngine

    @Before
    fun setUp() {
        engine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 50.0,
            initialAccuracyIndex = 0,
        )
    }

    @Test
    fun `first sample returns null (baseline)`() {
        assertNull(engine.processSample(0.85, 0), "First sample should be baseline, no adjustment")
    }

    @Test
    fun `sample too soon returns null`() {
        engine.processSample(0.85, 0)
        assertNull(engine.processSample(0.84, 5 * MINUTE_MS))
    }

    /**
     * The field failure, in the form the device produced it: two samples five
     * minutes apart across one reporting step, which the old engine read as a
     * 60 %/hr drain and acted on (#393).
     */
    @Test
    fun `a single reporting step does not throttle`() {
        engine.processSample(0.25, 0)
        assertNull(engine.processSample(0.24, 5 * MINUTE_MS))
        assertEquals(0, engine.throttleLevel, "throttled on one battery reporting step")
    }

    /**
     * A drain has to beat the budget by more than the window can resolve. Over
     * fifteen minutes one 1 % step is 4 %/hr, so a 5 %/hr budget cannot be
     * conclusively exceeded by a reading that close to it.
     */
    @Test
    fun `a drain inside the measurement resolution is inconclusive`() {
        engine.processSample(0.90, 0)
        assertNull(engine.processSample(0.88, 15 * MINUTE_MS))
        assertEquals(0, engine.throttleLevel)
    }

    @Test
    fun `no adjustment when within error threshold`() {
        engine.processSample(0.90, 0)
        // 5.04 %/hr against a 5 %/hr budget, over a window wide enough to
        // resolve it: on budget, so neither direction is conclusive.
        assertNull(engine.processSample(0.874, 60 * MINUTE_MS))
        assertEquals(0, engine.throttleLevel)
    }

    @Test
    fun `sustained heavy drain climbs one rung at a time`() {
        engine.processSample(0.90, 0)
        assertNull(
            engine.processSample(0.75, 30 * MINUTE_MS),
            "one conclusive window is not a dwell",
        )
        assertEquals(0, engine.throttleLevel)

        assertNotNull(engine.processSample(0.60, 60 * MINUTE_MS))
        assertEquals(1, engine.throttleLevel)

        assertNull(engine.processSample(0.45, 90 * MINUTE_MS))
        assertNotNull(engine.processSample(0.30, 120 * MINUTE_MS))
        assertEquals(2, engine.throttleLevel, "one rung per dwell, never two")
    }

    /**
     * The opt-out the old engine destroyed first: `0` meant "record every fix",
     * and `(0 * 1.5).clamp(10, 5000)` turned it into 10 m — permanently, since
     * the recovery path clamped at 10 too (#393).
     */
    @Test
    fun `a configured distance filter of zero is never clamped up`() {
        val zeroEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 0.0,
            initialAccuracyIndex = 0,
        )
        assertEquals(0.0, zeroEngine.distanceFilter)
        assertEquals(0.0, zeroEngine.throttleState.distanceFilter)
    }

    @Test
    fun `the ladder never goes below the configured values`() {
        val wideEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 250.0,
            initialAccuracyIndex = 2,
        )
        climbOneRung(wideEngine)
        assertEquals(1, wideEngine.throttleLevel)
        assertEquals(
            250.0,
            wideEngine.throttleState.distanceFilter,
            "a configured filter wider than the rung stands",
        )
        assertEquals(
            2,
            wideEngine.throttleState.desiredAccuracy,
            "a configured tier coarser than the rung stands",
        )
    }

    /**
     * Cadence before fidelity. On iOS the tier below `kCLLocationAccuracyBest`
     * is `…HundredMeters` — a hundredfold degradation in one step, and the step
     * the old engine took first. Against a 15 m tracking gate that guaranteed
     * every subsequent fix would be rejected (#396).
     */
    @Test
    fun `accuracy survives the first two rungs`() {
        climbOneRung(engine)
        assertEquals(1, engine.throttleLevel)
        assertEquals(0, engine.throttleState.desiredAccuracy)
        assertEquals(0, engine.throttleState.trackingAccuracyFloor)
        assertTrue(engine.throttleState.cadenceMultiplier > 1.0, "cadence is the first knob")
    }

    /** A coarsened request must relax the gate that judges its fixes. */
    @Test
    fun `a coarsened accuracy tier carries a matching gate floor`() {
        repeat(3) { climbOneRung(engine, offsetMinutes = it * 60L) }
        assertEquals(3, engine.throttleLevel)
        assertTrue(engine.throttleState.desiredAccuracy >= 1)
        assertTrue(
            engine.throttleState.trackingAccuracyFloor >= 100,
            "a 100 m accuracy tier behind a 15 m gate records nothing",
        )
    }

    @Test
    fun `charging lifts the throttle immediately`() {
        climbOneRung(engine)
        assertEquals(1, engine.throttleLevel)

        assertNotNull(engine.noteCharging(61 * MINUTE_MS))
        assertEquals(0, engine.throttleLevel)
        assertNull(
            engine.noteCharging(62 * MINUTE_MS),
            "no event when there was nothing to lift",
        )
    }

    @Test
    fun `accuracy index stays within 0-4`() {
        assertEquals(0, engine.accuracyIndex)
        repeat(6) { climbOneRung(engine, offsetMinutes = it * 60L) }
        assertTrue(engine.accuracyIndex in 0..4)
        assertTrue(engine.throttleLevel <= 4, "the ladder is bounded")
    }

    @Test
    fun `reset clears the measurement baseline`() {
        engine.processSample(0.90, 0)
        engine.reset()
        assertNull(
            engine.processSample(0.75, 30 * MINUTE_MS),
            "after reset, the next sample is a fresh baseline",
        )
    }

    @Test
    fun `periodic interval is stretched when present`() {
        val periodicEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 50.0,
            initialAccuracyIndex = 0,
            initialPeriodicInterval = 900,
        )
        val event = climbOneRung(periodicEngine)
        assertNotNull(event)
        assertNotNull(event.newPeriodicInterval)
        assertTrue(event.newPeriodicInterval!! > 900, "should stretch the periodic interval")
    }

    @Test
    fun `periodic interval stays null when not periodic`() {
        val event = climbOneRung(engine)
        assertNotNull(event)
        assertNull(event.newPeriodicInterval, "non-periodic engine should have null interval")
    }

    @Test
    fun `budget adjustment event contains correct target`() {
        val event = climbOneRung(engine)
        assertNotNull(event)
        assertEquals(5.0, event.targetBudget)
    }

    @Test
    fun `the engine reports the resolution behind its own drain figure`() {
        engine.processSample(0.90, 0)
        engine.processSample(0.75, 30 * MINUTE_MS)
        val state = engine.throttleState
        assertEquals(1_800.0, state.lastMeasurementSeconds)
        assertEquals(QUANTUM * 2, state.lastMeasurementResolution, absoluteTolerance = 1e-9)
        assertEquals(30.0, state.lastDrain, absoluteTolerance = 1e-9)
    }

    // =========================================================================
    // Helpers
    // =========================================================================

    /**
     * Two consecutive conclusive over-budget windows — the dwell one rung takes.
     *
     * 30 %/hr over half-hour windows, comfortably past the 2 %/hr such a window
     * can resolve on Android.
     */
    private fun climbOneRung(
        engine: BatteryBudgetEngine,
        offsetMinutes: Long = 0,
    ): BudgetAdjustmentEvent? {
        val base = offsetMinutes * MINUTE_MS
        engine.processSample(0.90, base)
        engine.processSample(0.75, base + 30 * MINUTE_MS)
        return engine.processSample(0.60, base + 60 * MINUTE_MS)
    }
}
